sql/parser/walk.go

```diff
@@ -0,0 +1,168 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package parser
+
+import "fmt"
+
+// The Visitor Visit method is invoked for each Expr node encountered by
+// WalkExpr. The returned Expr
+type Visitor interface {
+	Visit(expr Expr) Expr
+}
+
+// WalkExpr traverses the nodes in an expression. It starts by calling
+// v.Visit(expr). It then recursively traverses the children nodes of the
+// expression returned by v.Visit().
+func WalkExpr(v Visitor, expr Expr) Expr {
+	expr = v.Visit(expr)
+
+	switch t := expr.(type) {
+	case *AndExpr:
+		t.Left = WalkExpr(v, t.Left)
+		t.Right = WalkExpr(v, t.Right)
+
+	case *OrExpr:
+		t.Left = WalkExpr(v, t.Left)
+		t.Right = WalkExpr(v, t.Right)
+
+	case *NotExpr:
+		t.Expr = WalkExpr(v, t.Expr)
+
+	case *ParenExpr:
+		t.Expr = WalkExpr(v, t.Expr)
+
+	case *ComparisonExpr:
+		t.Left = WalkExpr(v, t.Left)
+		t.Right = WalkExpr(v, t.Right)
+
+	case *RangeCond:
+		t.Left = WalkExpr(v, t.Left)
+		t.From = WalkExpr(v, t.From)
+		t.To = WalkExpr(v, t.To)
+
+	case *NullCheck:
+		t.Expr = WalkExpr(v, t.Expr)
+
+	case *ExistsExpr:
+		// TODO(pmattis): Should we recurse into the Subquery?
+
+	case BytesVal:
+		// Terminal node: nothing to do.
+
+	case StrVal:
+		// Terminal node: nothing to do.
+
+	case IntVal:
+		// Terminal node: nothing to do.
+
+	case NumVal:
+		// Terminal node: nothing to do.
+
+	case BoolVal:
+		// Terminal node: nothing to do.
+
+	case ValArg:
+		// Terminal node: nothing to do.
+
+	case NullVal:
+		// Terminal node: nothing to do.
+
+	case QualifiedName:
+		// Terminal node: nothing to do.
+
+	case Tuple:
+		for i := range t {
+			t[i] = WalkExpr(v, t[i])
+		}
+
+	case Datum:
+		// Terminal node: nothing to do.
+
+	case *Subquery:
+		// TODO(pmattis): Should we recurse into the Subquery?
+
+	case *BinaryExpr:
+		t.Left = WalkExpr(v, t.Left)
+		t.Right = WalkExpr(v, t.Right)
+
+	case *UnaryExpr:
+		t.Expr = WalkExpr(v, t.Expr)
+
+	case *FuncExpr:
+		for i := range t.Exprs {
+			t.Exprs[i] = WalkExpr(v, t.Exprs[i])
+		}
+
+	case *CaseExpr:
+		if t.Expr != nil {
+			t.Expr = WalkExpr(v, t.Expr)
+		}
+		for _, w := range t.Whens {
+			w.Cond = WalkExpr(v, w.Cond)
+			w.Val = WalkExpr(v, w.Val)
+		}
+		if t.Else != nil {
+			t.Else = WalkExpr(v, t.Else)
+		}
+
+	case *CastExpr:
+		t.Expr = WalkExpr(v, t.Expr)
+
+	default:
+		panic(fmt.Sprintf("unsupported expression type: %T", expr))
+	}
+
+	return expr
+}
+
+// Args defines the interface for retrieving arguments. Return false for the
+// second return value if the argument cannot be found.
+type Args interface {
+	Arg(i int) (Datum, bool)
+}
+
+type argVisitor struct {
+	args Args
+	err  error
```

Comment:
Right, I don't know how this code is going to grow, but it's something to be aware of. Deviating from standard idioms like this can make future refactorings problematic.